### PR TITLE
Adding channel padding support for blazeface model.

### PR DESCRIPTION
**Pull request by Timen**
Important changes:
1. added support for channel padding in the padding layer
2. added arm and vulkan optimization support (reflective and edge channel padding is limited to elempack =1)
3. added blazeface model conversion support
4. added blazeface model to the benchmark.

### CODE REVIEW

**Comment by tencent-adm** [comment]
[CLA assistant check](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1826) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.tencent.com/Tencent/ncnn?pullRequest=1826) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.tencent.com/check/Tencent/ncnn?pullRequest=1826) it.</sub>

**Comment by Timen** [comment]
Some build errors seem to be occurring:
```

W: GPG error: https://repo.mongodb.org/apt/ubuntu bionic/mongodb-org/4.2 Release: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 656408E390CFB1F5
E: The repository 'https://repo.mongodb.org/apt/ubuntu bionic/mongodb-org/4.2 Release' is not signed.
The command "eval "${BEFORE_BUILD}"" failed and exited with 100 during .
```

File: src/layer/arm/padding_arm.cpp [hunk 1]
```diff
@@ -150,10 +175,32 @@ int Padding_arm::forward_bf16s(const Mat& bottom_blob, Mat& top_blob, const Opti
...
+            //Reflective padding and edge padding only supports channel padding in elempack 1
+            out_elempack = 1;
+        }
+        outc = padded_channels/out_elempack;
+    }
+    if (out_elempack != elempack){
+            Option opt_pack = opt;
+            opt_pack.blob_allocator = opt.workspace_allocator;
+            convert_packing(bottom_blob, bottom_blob_unpacked, 1, opt_pack);
+    }
```
Comment (nihui) [inline src/layer/arm/padding_arm.cpp:200]: Move this block into the upper "if (front != 0 || behind != 0 )" block

File: src/layer/arm/padding_arm.cpp [hunk 2]
```diff
@@ -70,15 +70,37 @@ int Padding_arm::forward(const Mat& bottom_blob, Mat& top_blob, const Option& op
...
```
Comment (nihui) [inline src/layer/arm/padding_arm.cpp:95]: Move this block into the upper "if (front != 0 || behind != 0 )" block

File: src/layer/arm/padding_arm.cpp [hunk 3]
```diff
@@ -70,15 +70,37 @@ int Padding_arm::forward(const Mat& bottom_blob, Mat& top_blob, const Option& op
...
+    if (elempack == 4 && out_elempack == 4)
     {
+
         int outw = w + left + right;
 
         if (dims == 1)
         {
-            top_blob.create(outw, elemsize, elempack, opt.blob_allocator);
+            top_blob.create(outw, elemsize, out_elempack, opt.blob_allocator);
```
Comment (nihui) [inline src/layer/arm/padding_arm.cpp:103]: out_elempack must be equal to elempack here
prefer elempack

File: src/layer/arm/padding_arm.cpp [hunk 4]
```diff
@@ -96,7 +118,7 @@ int Padding_arm::forward(const Mat& bottom_blob, Mat& top_blob, const Option& op
 
         if (dims == 2)
         {
-            top_blob.create(outw, outh, elemsize, elempack, opt.blob_allocator);
+            top_blob.create(outw, outh, elemsize, out_elempack, opt.blob_allocator);
```
Comment (nihui) [inline src/layer/arm/padding_arm.cpp:121]: out_elempack must be equal to elempack here
prefer elempack

File: src/layer/arm/padding_arm.cpp [hunk 5]
```diff
@@ -112,24 +134,28 @@ int Padding_arm::forward(const Mat& bottom_blob, Mat& top_blob, const Option& op
 
         if (dims == 3)
         {
-            top_blob.create(outw, outh, channels, elemsize, elempack, opt.blob_allocator);
+            top_blob.create(outw, outh, outc, elemsize, out_elempack, opt.blob_allocator);
             if (top_blob.empty())
                 return -100;
-
+            int front_ = front/out_elempack;
```
Comment (nihui) [inline src/layer/arm/padding_arm.cpp:140]: out_elempack must be equal to elempack here
prefer elempack

File: src/layer/vulkan/padding_vulkan.cpp [hunk 6]
```diff
@@ -224,21 +268,29 @@ int Padding_vulkan::forward(const std::vector<VkMat>& bottom_blobs, std::vector<
...
     {
         const int* param_data = reference_blob.mapped();
 
         _top = param_data[0];
         _bottom = param_data[1];
         _left = param_data[2];
         _right = param_data[3];
+        if (reference_blob.total() >= 9){
+            _front = param_data[7];
+            _behind = param_data[8];
```
Comment (nihui) [inline src/layer/vulkan/padding_vulkan.cpp:286]: the reference_blob contains the padding border constants only.
use param_data[4] and param_data[5] for front and behind param

File: src/layer/vulkan/padding_vulkan.cpp [hunk 7]
```diff
@@ -182,39 +195,70 @@ int Padding_vulkan::forward(const VkMat& bottom_blob, VkMat& top_blob, VkCompute
...
+    {
+        if (out_elempack == 8) out_elemsize = 8*2u;
+        if (out_elempack == 4) out_elemsize = 4*2u;
+        if (out_elempack == 1) out_elemsize = 4u;
+    }
+    // unpacking
+    VkMat bottom_blob_unpacked = bottom_blob;
+    if (elempack != out_elempack)
+    {
+        vkdev->convert_packing(bottom_blob, bottom_blob_unpacked, out_elempack, cmd, opt);
```
Comment (nihui) [inline src/layer/vulkan/padding_vulkan.cpp:226]:
```
Opt opt_unpack = opt;
opt_unpack.blob_vkallocator = opt.workspace_vkallocator;
```

File: src/layer/vulkan/padding_vulkan.cpp [hunk 8]
```diff
@@ -249,47 +301,76 @@ int Padding_vulkan::forward(const std::vector<VkMat>& bottom_blobs, std::vector<
...
```
Comment (nihui) [inline src/layer/vulkan/padding_vulkan.cpp:332]:
```
Opt opt_unpack = opt;
opt_unpack.blob_vkallocator = opt.workspace_vkallocator;
```

File: src/layer/vulkan/padding_vulkan.cpp [hunk 9]
```diff
@@ -302,39 +383,71 @@ int Padding_vulkan::forward(const VkImageMat& bottom_blob, VkImageMat& top_blob,
...
+    {
+        if (out_elempack == 8) out_elemsize = 8*2u;
+        if (out_elempack == 4) out_elemsize = 4*2u;
+        if (out_elempack == 1) out_elemsize = 4u;
+    }
+    // unpacking
+    VkImageMat bottom_blob_unpacked = bottom_blob;
+    if (elempack != out_elempack)
+    {
+        vkdev->convert_packing(bottom_blob, bottom_blob_unpacked, out_elempack, cmd, opt);
```
Comment (nihui) [inline src/layer/vulkan/padding_vulkan.cpp:414]:
```
Opt opt_unpack = opt;
opt_unpack.blob_vkallocator = opt.workspace_vkallocator;
```

File: src/layer/vulkan/padding_vulkan.cpp [hunk 10]
```diff
@@ -369,37 +491,65 @@ int Padding_vulkan::forward(const std::vector<VkImageMat>& bottom_blobs, std::ve
...
```
Comment (nihui) [inline src/layer/vulkan/padding_vulkan.cpp:522]:
```
Opt opt_unpack = opt;
opt_unpack.blob_vkallocator = opt.workspace_vkallocator;
```

File: src/layer/vulkan/padding_vulkan.cpp [hunk 11]
```diff
@@ -349,16 +462,25 @@ int Padding_vulkan::forward(const std::vector<VkImageMat>& bottom_blobs, std::ve
...
     {
         const int* param_data = reference_blob.mapped();
 
         _top = param_data[0];
         _bottom = param_data[1];
         _left = param_data[2];
         _right = param_data[3];
+        if (reference_blob.total() >= 9){
+            _front = param_data[7];
+            _behind = param_data[8];
```
Comment (nihui) [inline src/layer/vulkan/padding_vulkan.cpp:476]: the reference_blob contains the padding border constants only.
use param_data[4] and param_data[5] for front and behind param

File: tests/test_padding.cpp [hunk 12]
```diff
@@ -50,14 +54,24 @@ static int test_padding(int w, int h, int c, int top, int bottom, int left, int
...
+        || test_padding(5, 7, 8, 2, 3, 2, 3, 1, 0.f, 22,8,0)
+        || test_padding(5, 7, 6, 4, 3, 4, 3, 2, 0.f, 0,2,4)
+        || test_padding(5, 7, 7, 0, 1, 0, 1, 0, 233.f, 0,3,1)
+        || test_padding(5, 7, 3, 2, 1, 2, 1, 0, 0.f, 3,0,4)
+        || test_padding(5, 7, 3, 2, 1, 2, 1, 0, 0.f, 3,0,4)
+        || test_padding(5, 7, 3, 2, 1, 2, 1, 0, 0.f, 3,0,4)
+        || test_padding(5, 7, 16, 2, 1, 2, 1, 2, 0.f, 3,4,4)
+        || test_padding(5, 7, 16, 2, 1, 2, 1, 0, 0.f, 3,8,8)
+        || test_padding(5, 7, 16, 2, 1, 2, 1, 1, 0.f, 3,8,8)
+        || test_padding(5, 7, 16, 2, 1, 2, 1, 2, 0.f, 3,8,8)
```
Comment (nihui) [inline tests/test_padding.cpp:74]: duplicated test cases

**Comment by nihui** [comment]
Thanks !

**Pull request merged**